In [63]:
import numpy as np
import matplotlip.pyplot as plt

ImportError: No module named matplotlip.pyplot

In [64]:
def mag(vec):
    return np.dot(vec,vec)**.5

**1) De verctor de estado a elementos orbitales** Un satelite en  ́orbita alrededor de la Tierra
tiene una posicion instantanea $\vec{r} = 6045,0i+ 3490,0ˆj + 0ˆk [km]$ y una velocidad $\vec{v} = −2,457ˆi+
6,618ˆj+2,533ˆk [km/s]$. Determine los elementos orbitales suponiendo que el plano fundamental
es el ecuador terrestre.

In [65]:
r=np.array([6045.0,3490,0])
v=np.array([-2.457,6.618,2.533])

In [66]:
i=np.array([1,1,0])
j=np.array([1,1,0])
k=np.array([0,0,1])

$$\vec{n}=\hat{k}\times\vec{h}$$
$$\vec{h}=\vec{r} \times \dot{ \vec{r}}$$


In [67]:
h=np.cross(r,v)
n=np.cross(k,h)

Como es un satélite que orbita la tierra supondremos que $M_{satelite}<<M_{tierra}$. Así, $\mu=G(M_t+M_s)\approx GM_t$.


$$\mu=3.983\times 10^{24}m^3/s^2$$

$$\vec{A}=\dot{\vec{r}}\times\vec{h}-\mu\frac{\vec{r}}{r}$$

$$ \vec{e}=\frac{\vec{A}}{\mu}$$

In [68]:
mu=3.983e24
A=np.cross(v,h)-mu*r/mag(r)**.5
e=A/mu

$$\cos(\Omega)=\frac{\vec{n} . \hat{i}}{n}$$

$$\cos\omega=\frac{\vec{n}.\vec{e}}{ne}$$

$$\cos f=\frac{\vec{e}\vec{r}}{er}$$

$$\cos i= \frac{\vec{h}.\hat{k}}{h}$$

In [69]:
if np.dot(n,j)>0:
    Omega=np.arccos(np.dot(n,i)/mag(n))*180./np.pi
else: 
    Omega=2.*np.pi-np.arccos(np.dot(n,i)/mag(n))*180./np.pi
    
if np.dot(e,k)>=0:
    omega=np.arccos(np.dot(n,e)/(mag(n)*mag(e)))*180./np.pi
else: 
    omega=2.*np.pi-np.arccos(np.dot(n,e)/(mag(n)*mag(e)))*180/np.pi
    
vr= np.dot(r,v)/mag(r)

if vr>=0:
    f=np.arccos(np.dot(e,r)/(mag(e)*mag(r)))*180./np.pi
else:
    f=2*np.pi-np.arccos(np.dot(e,r)/(mag(e)*mag(r)))*180./np.pi
    
ii=np.arccos(np.dot(h,k)/mag(h))*180./np.pi

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: invalid value encountered in arccos
  
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:14: RuntimeWarning: invalid value encountered in arccos
  


## EL ARGUMENTO DEL COSENO ES MAYOR QUE 1 ¿QUÉ SE HACE?

**2). Satelite en  ́orbita.**  Un sat ́elite en  ́orbita elıptica alrededor de la Tierra tiene un perıodo de
15.7430 horas y una distancia al perigeo de 12756.0 km (respecto al centro de la Tierra). En
t = 10 horas despues del paso por el perigeo calcule:

In [70]:
T=15.7430*60**2
G=6.67e-11
d=12756000 #Distanica ap perigeo
M_t=5.97e24 #Masa tierra
t=10*60**2 

Por tercera ley de Kepler, $$a=\Big(\frac{T^2\mu}{4\pi^2}\Big)^{1/3}$$.

Además $d$ la distancia al perigeo, es $d=a-c$.

$$e=\frac{c}{a} $$

In [71]:
a=(T**2*G*M_t/(4.*np.pi**2))**(1./3)
c=a-d
e=c/a
f_0=0 #Anomalía verdadera en t=0
p=a*(1-e**2) #Semilatus Rectum
r_0=p/(1+np.cos(f_0)) #Distancia radial en t=0

e

0.5998639287851092

Como a $e=0.59986<0.6627434$, se usa la solución a la ecuación de Kepler:
$$E=M+sin(E)$$

Donde $M=n(t-t_p)$, es la anmalía media $t_p$ el tiempo de paso por el perigeo y $n=\frac{2\pi}{T}$ el movimiento medio 

In [72]:
h=(p*mu)**.5

n=2*np.pi/T # Movimiento medio
t_p=0 #tiempo de paso por el perigeo

M=n*(t-t_p) #Anomalía media 

E=M
for i in range(100):
    E=M+e*np.sin(E)
    
E

3.6823147471140802

La ecuación converge a $E=3.6823$, así:

In [73]:
r_t=a*(1-e*np.cos(E)) #Distancia radial en t=10 horas

r_t/1000

48274.14799431139

**a)** La distancia radial luego de diez horas del paso por el perigeo es 48274 km

Se sabe que $$\dot{r}=\frac{h}{p}e sin f$$
y además $$cos f=\frac{-ecosE}{1-ecosE}$$

In [74]:
f=np.arccos(-e*np.cos(E)/(1-e*np.cos(E)))
v=h/p*e*np.sin(f) 
 

## ESTA NO ES LA COMPONENTE RADIAL ¿CÓMO SE SACA?

Si la tasa de cambio de la variable radial $V_r=\frac{\vec{r}\vec{v}}{r}$, es menor que cero, se acerca al periapsis.

**8) Puntos colineales de Lagrange** [numerico]. Calcule numericamente la ubicacion de los
tres puntos colineales de Lagrange para un valor de α = 0,3 resolviendo la ecuacion.

$$x-(1-\alpha)\frac{x-x_1}{|x-x_1|^3}-\alpha\frac{x-x_1}{|x-x_2|^3}=0$$

In [75]:
alpha=.3

In [76]:
#Supndré las posiciónes de los dos cuerpos como:
x1=-1
x2=2

**Usando el método de cruce de recatas, se encuentran los puntos de lagrange.**

In [77]:
rango=[-20,20]
def f1(x): 
    return x-(1-alpha)/abs(x-x1)**3-alpha/abs(x-x2)**3
def f2(x):
    return 0

x_array=np.arange(rango[0],rango[1],.001)
y1_array=[f1(i) for i in x_array]
y2_array=[f2(i) for i in x_array]

ptosx=[]
ptosy1=[]
ptosy2=[]

# Verificamos que haya interseccion
for i in range(len(x_array)-1):
    if y1_array[i]<y2_array[i] and y1_array[i+1]>=y2_array[i+1] or y1_array[i]>=y2_array[i] and y1_array[i+1]<y2_array[i+1]:
    
        ptosx.append(x_array[i])
        ptosx.append(x_array[i+1])
        
        ptosy1.append(y1_array[i])
        ptosy1.append(y1_array[i+1])
    
        ptosy2.append(y2_array[i])
        ptosy2.append(y2_array[i+1])


def cruce_rectas(x1,x2,y11,y12,y21,y22):
    
    m1=(y12-y11)/(x2-x1)
    m2=(y22-y21)/(x2-x1)
    
    if (m1!=m2):
        
        b1=-x1*m1+y11
        b2=-x1*m2+y21

        #Pto de interseccion (a,c)
        a=(b2-b1)/(m1-m2)
        c=a*m1+b1
    
        return (a,c)

#Se recorre la lista x_array usando la funcion cruce_rectas 
ptos_crucex=[]
ptos_crucey=[]

#El indice es por que cruce_rectas entrega dos argumentos
for i in range(0,len(ptosx),2):
    ptos_crucex.append(cruce_rectas(ptosx[i],ptosx[i+1],ptosy1[i],ptosy1[i+1],ptosy2[i],ptosy2[i+1])[0])
    ptos_crucey.append(cruce_rectas(ptosx[i],ptosx[i+1],ptosy1[i],ptosy1[i+1],ptosy2[i],ptosy2[i+1])[1])

print("Se encontraron %d intersecciones en el rango dado \n" %(len(ptos_crucex)))
for i in range(len(ptos_crucex)):
    print (ptos_crucex[i],ptos_crucey[i])

Se encontraron 3 intersecciones en el rango dado 

(0.35085644985835357, 0.0)
(1.3930943883071873, 0.0)
(2.4946801271364145, 0.0)


**Los punto de Lagrange encontrados para dos cuerpos con posiciones $x_1=-1$ y $x_2=2$ son:**

$$x_{L1}=0.3509$$ $$x_{L2}=1.3931$$ $$x_{L3}=2.4947$$ 